zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,294 @@
+//! Sapling note and value commitments and types.
+
+#![allow(clippy::unit_arg)]
+
+#[cfg(test)]
+mod arbitrary;
+
+use std::{fmt, io};
+
+use bitvec::prelude::*;
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{
+    keys::sapling::{find_group_hash, Diversifier, TransmissionKey},
+    serde_helpers,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+    types::amount::{Amount, NonNegative},
+};
+
+/// Generates a random scalar from the scalar field \mathbb{F}_r_𝕁.
+///
+/// The prime order subgroup 𝕁^(r) is the order-r_𝕁 subgroup of 𝕁
+/// after the Edwards cofactor h_𝕁 = 8 is factored out. This function
+/// is useful when generating the uniform distribution on
+/// \mathbb{F}_r_𝕁 needed for Sapling commitment schemes' trapdoor
+/// generators.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#jubjub
+pub fn generate_trapdoor<T>(csprng: &mut T) -> jubjub::Fr
+where
+    T: RngCore + CryptoRng,
+{
+    let mut bytes = [0u8; 32];
+    csprng.fill_bytes(&mut bytes);
+    jubjub::Fr::from_bytes(&bytes).unwrap()
```

Comment:
I think the unwrap will panic if the scalar representation is out of range: https://github.com/zcash/librustzcash/blob/f55f094ef69c2caeaa941bcf851dac8cb5fa9942/jubjub/src/fr.rs#L210-L237

You need an unbiased scalar; I would recommend doing it in the way described for RedDSA.GenRandom in the protocol spec.

Reply:
What about just a mul mod r, similar to: 
https://github.com/zcash/librustzcash/blob/7b94622ae0dad94f2720e31c976838ee922677c7/zcash_primitives/src/jubjub/fs.rs#L613

Reply:
That will work. The original motivation for how GenRandom is defined was to tolerate potential biases in the RNG. That probably isn't too much of a problem if you're careful which RNG is used.